### PR TITLE
Generate array shapes for PHPStan

### LINKED ISSUE #15: Implement short inline syntax for array parameters and return values
Neither PHPStan nor Psalm support the hash-style notation used by WordPress for its array parameters and array return values.

```
@param array $args {
    Array of arguments used to display the help tab.

    @type string   $title    Title for the tab. Default false.
    @type string   $id       Tab ID. Must be HTML-safe and should be unique for this menu.
                             It is NOT allowed to contain any empty spaces. Default false.
    @type string   $content  Optional. Help tab content in plain text or HTML. Default empty string.
    @type callable $callback Optional. A callback to generate the tab content. Default false.
    @type int      $priority Optional. The priority of the tab, used for ordering. Default 10.
}
```

This means there's currently no way to inform static analysis tools of the types of the array elements passed to or returned by functions such as `wp_remote_post()` and classes such as `WP_Query`.

It would be great to:

1. Post-process the stubs after they're generated and convert hash-style array notation in the docblocks into short inline syntax. Example:
    ```diff
    - * @param array $args {
    - *     Array of arguments used to display the help tab.
    - *
    - *     @type string   $title    Title for the tab. Default false.
    - *     @type string   $id       Tab ID. Must be HTML-safe and should be unique for this menu.
    - *                              It is NOT allowed to contain any empty spaces. Default false.
    - *     @type string   $content  Optional. Help tab content in plain text or HTML. Default empty string.
    - *     @type callable $callback Optional. A callback to generate the tab content. Default false.
    - *     @type int      $priority Optional. The priority of the tab, used for ordering. Default 10.
    - * }
    + * @param array{
    + *     title?: string,
    + *     id?: string,
    + *     content?: string,
    + *     callback?: callable,
    + *     priority?: int,
    + * } $args
    ```
2. Create a mapping from function parameters such as `wp_remote_post() $args` to their canonical function parameter such as `WP_Http::request() $args` and then apply the same docblock from the canonical function in its wrapper functions. This allows for parameter type checking in wrapper functions which are more commonly used. This likely needs to happen as a separate second step as it will be a manual process to create the mapping. WordPress core isn't well documented or consistently documented in this regard.

## Approach

It's possible that [`johnbillion/wp-parser-lib`](https://github.com/johnbillion/wp-parser-lib) could be used to aid the transformation as it's able to tokenise the hash style array notation.

## Considerations

Unfortunately the short inline syntax supported by PHPStan and Psalm don't allow for human-readable documentation of the array elements. This means the stubs will be better for PHPStan and Psalm but worse for humans. Maybe two sets of stubs needs to be created, one with the human-readable docblock from core and one with the short inline syntax for use in static analysis.

## Refs

* [PHPStan array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes)
* [Psalm array shapes](https://psalm.dev/docs/annotating_code/type_syntax/array_types/#object-like-arrays)

### PR DESCRIPTION
**Pull request by johnbillion**
Work in progress. Fixes #15.

This converts the array hash notation that WordPress uses into an array shape that PHPStan understands, and appends a `@phpstan-param` tag for it to the docblock of the generated stub. It's fault tolerant so any incorrectly formatted array hashes should simply be ignored.

Todo:

* [x] `@param`
* [x] `@return`
* [x] Investigate whether elements in `@phpstan-return` tags should be optional
* [x] Investigate handling elements in `@param` marked as "Required"

I tried to add handling for "Required" and "Optional" elements but core isn't consistent enough with its documentation and there aren't many anyway. Not worth worrying about in my opinion.

When the stubs are generated this adds:

* 231 `@phpstan-param` tags
* 65 `@phpstan-return` tags

Example:

```
* @param array|string $args {
*     Optional. Array or string of arguments for installing a package. Default empty array.
*
*     @type string $source                      Required path to the package source. Default empty.
*     @type string $destination                 Required path to a folder to install the package in.
*                                               Default empty.
*     @type bool   $clear_destination           Whether to delete any files already in the destination
*                                               folder. Default false.
*     @type bool   $clear_working               Whether to delete the files form the working directory
*                                               after copying to the destination. Default false.
*     @type bool   $abort_if_destination_exists Whether to abort the installation if
*                                               the destination folder already exists. Default true.
*     @type array  $hook_extra                  Extra arguments to pass to the filter hooks called by
*                                               WP_Upgrader::install_package(). Default empty array.
* }
* @phpstan-param array{
*   source?: string,
*   destination?: string,
*   clear_destination?: bool,
*   clear_working?: bool,
*   abort_if_destination_exists?: bool,
*   hook_extra?: array,
* } $args
```

### CODE REVIEW

**Comment by szepeviktor** [comment]
Thank you for your work, John!
Please add PHPStan to test this code.

**Comment by johnbillion** [comment]
I've done that locally as part of my testing but there are several errors that need ignoring. I'll carry on with this later this week.

File: phpstan.neon [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        - '#duplicate keys with value#'
+        - '#WP_HTTP_Streams#'
+        - '#unused parameter#'
+        - '#Constant [A-Z0-9_]+ not found#'
+        - '#Default value of the parameter#'
+        - '#\stream#'
+        - '#\GdImage#'
+        - '#\IRI#'
+        - '#Requests#'
+        - '#SimplePie#'
```
Comment (szepeviktor) [inline phpstan.neon:17]: Oh nooo!
```suggestion
    level: 7
    paths:
        - finder.php
        - visitor.php
```

**Comment by szepeviktor** [comment]
> there are several errors that need ignoring.

That was a mistake. Check only working PHP source code :)

**Comment by johnbillion** [comment]
Ah sorry what I meant was I was using PHPStan to validate the generated array shapes (to check to syntax errors etc) and as part of that testing I had to ignore other unrelated errors.

**Comment by szepeviktor** [comment]
> using PHPStan to validate the generated array shapes

That is a valid point but there is no way in PHPStan to pick-and-choose source code parts.

**Comment by szepeviktor** [comment]
This PR is about we going to WordPress but WordPress should come to its senses!

**Comment by johnbillion** [comment]
Yeah I would prefer it if WordPress used a more widely used syntax but its array hash notation does have one advantage over the syntax used by PHPStan in that each array element supports a description which is very useful when reading the code.

**Comment by szepeviktor** [comment]
> each array element supports a description

I knew you support noobs!!

**Comment by szepeviktor** [comment]
Green to go 🟢 now that we have https://github.com/php-stubs/generator/pull/5

File: visitor.php [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        if (strpos($text, '    @type ') === false) {
+            return null;
+        }
+
+        // Populate `$types` with the value of each top level `@type`.
+        $types = preg_split('/\R+    @type /', $text);
+
+        if (!$types) {
+            return null;
+        }
```
Comment (szepeviktor) [inline visitor.php:208]: Please avoid type juggling. `if` needs a boolean.
```suggestion
        if ($types === false) {
            return null;
        }
```
https://www.php.net/manual/en/function.preg-split.php

File: wordpress-stubs.php [hunk 3]
```diff
@@ -676,6 +684,15 @@ public function install_package($args = array())
...
          *                              or false if unable to connect to the filesystem.
+         * @phpstan-param array{
+         *   package?: string,
+         *   destination?: string,
+         *   clear_destination?: bool,
+         *   clear_working?: bool,
+         *   abort_if_destination_exists?: bool,
+         *   is_multi?: bool,
+         *   hook_extra?: array,
+         * } $options
```
Comment (szepeviktor) [inline wordpress-stubs.php:695]: How about union types?
What will we do about them?

File: visitor.php [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+        unset($types[0]);
+        $elements = [];
+
+        foreach ($types as $typeTag) {
+            $parts = preg_split('#\s+#', trim($typeTag));
+
+            if (!$parts || count($parts) < 2) {
+                return null;
+            }
```
Comment (szepeviktor) [inline visitor.php:218]:
```suggestion
            if ($parts === false || count($parts) < 2) {
                return null;
            }
```
https://www.php.net/manual/en/function.preg-split.php

**Comment by szepeviktor** [comment]
Thank you John for your work. I think it is ready to merge.
It'll be merged when WP core 5.9 will be released.

Have you tested some of your projects with these array shapes?

**Comment by johnbillion** [comment]
Yeah I've been testing this, mostly the functions with new `@phpstan-param` tags. As all their array elements are optional it just helps with incorrect types and not much else, so these are safe to add.

I should do some more testing of the functions that now include `@phpstan-return` tags because I've not tested those much.

**Comment by johnbillion** [comment]
I checked all the `@phpstan-return` tags and they're all good 👍

**Pull request merged**